src/tools/db-upgrade/DbUpgrader.cpp

```diff
@@ -0,0 +1,461 @@
+                data.emplace_back(std::move(newIndexKey), "");
+            }
+        }
+    }
+}
+
+std::string DbUpgrader::indexVertexKey(PartitionID partId,
+                VertexID& vId,
+                RowReader* reader,
+                std::shared_ptr<nebula::meta::cpp2::IndexItem> index) {
+    auto values = IndexKeyUtils::collectIndexValues(reader, index->get_fields());
+    if (!values.ok()) {
+        return "";
+    }
+    return IndexKeyUtils::vertexIndexKey(spaceVidLen_,
+                                         partId,
+                                         index->get_index_id(),
+                                         vId,
+                                         std::move(values).value());
+}
+
+// Think that the old and new values are exactly the same, so use one reader
+void DbUpgrader::encodeEdgeValue(PartitionID partId,
+                               RowReader* reader,
+                               const meta::NebulaSchemaProvider* schema,
+                               std::string& newkey,
+                               VertexID& svId,
+                               EdgeType type,
+                               EdgeRanking rank,
+                               VertexID& dstId,
+                               std::vector<kvstore::KV>& data) {
+    if (reader == nullptr) {
+        return;
+    }
+    auto ret = encodeRowVal(reader, schema);
+    if (ret.empty()) {
+        return;
+    }
+    data.emplace_back(std::move(newkey), std::move(ret));
+
+    if (type <= 0) {
+        return;
+    }
+
+    // encode v2 index value
+    auto it = edgeIndexes_.find(type);
+    if (it != edgeIndexes_.end()) {
+        for (auto& index : it->second) {
+            auto newIndexKey = indexEdgeKey(partId, reader, svId, rank, dstId, index);
+            if (!newIndexKey.empty()) {
+                data.emplace_back(std::move(newIndexKey), "");
+            }
+        }
+    }
+}
+
+std::string DbUpgrader::indexEdgeKey(PartitionID partId,
+                        RowReader* reader,
+                        VertexID& svId,
+                        EdgeRanking rank,
+                        VertexID& dstId,
+                        std::shared_ptr<nebula::meta::cpp2::IndexItem> index) {
+    auto values = IndexKeyUtils::collectIndexValues(reader, index->get_fields());
+    if (!values.ok()) {
+        return "";
+    }
+    return IndexKeyUtils::edgeIndexKey(spaceVidLen_,
+                                       partId,
+                                       index->get_index_id(),
+                                       svId,
+                                       rank,
+                                       dstId,
+                                       std::move(values).value());
+}
+
+void DbUpgrader::doProcessAllTagsAndEdges() {
+    auto ret = buildSchemaAndIndex();
+    if (!ret.ok()) {
+         LOG(INFO) << "Build schema and index failed";
+         return;
+    }
+
+    for (auto& partId : parts_) {
```

Comment:
We'd better print something when we start to process a new part (spaceId and partId is sufficient)

Reply:
good ！